lib/Optimizer/Transforms/ReplaceStateWithKernel.cpp

```diff
@@ -0,0 +1,129 @@
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+
+namespace cudaq::opt {
+#define GEN_PASS_DEF_REPLACESTATEWITHKERNEL
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+#define DEBUG_TYPE "replace-state-with-kernel"
+
+using namespace mlir;
+
+namespace {
+// clang-format off
+/// Replace `quake.get_number_of_qubits` by a call to a function
+/// that computes the number of qubits for a state.
+///
+/// ```mlir
+///  %0 = quake.materialize_state @callee.num_qubits_0 @callee.init_0 : !cc.ptr<!cc.state>
+///  %1 = quake.get_number_of_qubits %0 : (!cc.ptr<!cc.state>) -> i64
+/// ───────────────────────────────────────────
+///  %1 = call @callee.num_qubits_0() : () -> i64
+/// ```
+// clang-format on
+class ReplaceGetNumQubitsPattern
+    : public OpRewritePattern<quake::GetNumberOfQubitsOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(quake::GetNumberOfQubitsOp numQubits,
+                                PatternRewriter &rewriter) const override {
+
+    auto stateOp = numQubits.getOperand();
+    auto materializeState = stateOp.getDefiningOp<quake::MaterializeStateOp>();
+    if (!materializeState)
+      return numQubits->emitError(
+          "ReplaceStateWithKernel: failed to replace `quake.get_num_qubits`");
+
+    auto numQubitsFunc = materializeState.getNumQubitsFunc();
+    rewriter.setInsertionPoint(numQubits);
+    rewriter.replaceOpWithNewOp<func::CallOp>(
+        numQubits, numQubits.getType(), numQubitsFunc, mlir::ValueRange{});
+    return success();
+  }
+};
+
+// clang-format off
+/// Replace `quake.init_state` by a call to a (modified) kernel that produced
+/// the state.
+///
+/// ```mlir
+///  %0 = quake.materialize_state @callee.num_qubits_0 @callee.init_0 : !cc.ptr<!cc.state>
+///  %3 = quake.init_state %2, %0 : (!quake.veq<?>, !cc.ptr<!cc.state>) -> !quake.veq<?>
+/// ───────────────────────────────────────────
+/// %3 = call @callee.init_0(%2): (!quake.veq<?>) -> !quake.veq<?>
+/// ```
+// clang-format on
+class ReplaceInitStatePattern
+    : public OpRewritePattern<quake::InitializeStateOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(quake::InitializeStateOp initState,
+                                PatternRewriter &rewriter) const override {
+    auto allocaOp = initState.getOperand(0);
+    auto stateOp = initState.getOperand(1);
+
+    if (auto ptrTy = dyn_cast<cudaq::cc::PointerType>(stateOp.getType())) {
+      if (isa<cudaq::cc::StateType>(ptrTy.getElementType())) {
+        auto materializeState =
+            stateOp.getDefiningOp<quake::MaterializeStateOp>();
+        if (!materializeState)
+          return initState->emitError(
+              "ReplaceStateWithKernel: failed to replace `quake.init_state`");
```

Comment:
Again, `return failure();` ought to be enough. The codegen isn't going to generate code ultimately if nothing cleans this up.

That said, adding a `LLVM_DEBUG(llvm::dbgs() << "Whoops!\n");` line would help with debugging the pass.